Move methods of SQL backends from Client to Backend

datapythonista opened the pull request:
xref #3017

The diff is big, but the changes here are quite simple:
- Moving all methods of SQL backends from `Client` to `Backend`
- Creating proxy methods in `Client`, so the `Backend` methods are called
- Replacing `self.client` by `self` where needed (since attributes of `Backend` can now be called from `Backend` directly)
- Updates to tests. Mainly `MockConnection` and how tests are ignored when `create_table` doesn't exist.

jreback commented on ibis/backends/base/client.py:
```diff
@@ -29,6 +29,68 @@ def exists_database(self, name):
     def exists_table(self, name, database=None):
         return self.backend.exists_table(name, database)
 
+    # Non-standard methods
```
make the comment that these are for compatiblity, ref an issue here (e.g. to remove later in time)

jreback replied:
ideally if you can have the same signature; the doc-string not so important

jreback replied:
event better you could put these in a separate mixin class e.g. ClientDispatchToBackend or ClientCompat or something and just mix here. (then its also easy for a Backend to opt *out* of compatiblity once they are good.

jreback commented on ibis/backends/base/client.py:
```diff
@@ -29,6 +29,68 @@ def exists_database(self, name):
+        return self.backend.create_table(*args, **kwargs)
+
+    def drop_table(self, *args, **kwargs):
+        if not hasattr(self.backend, 'drop_table'):
```
why is this needed, shouldn't this notimplemented exist on the BaseBackend?

jreback reviewed (changes requested):
cconceptually +1 on this way. suggestions in line

datapythonista commented:
Thanks for the feedback @jreback. Good points, addressed your comments.  I don't think the mixin class is a good idea at this point. The methods there aren't specific to a backend (yet). Most of them should be standard probably, it's just that we didn't have the time to decide what should be in the common backend API, standardize their signatures and create the abstract methods. I think it'll make things easier for the future to keep all methods as they are, and when the API is fully defined, consider if we want to move the methods in the standard API to mixin classes. I created the issue #3019 and linked it in the comment to have this information there. Does it make sense?

datapythonista commented:
@jreback @cpcloud this is finally green. Sorry for the big diff, but I think with this we're almost ready for the release. We should open a similar PR for the pandas based backends, and no backend in this repo should implement much in the client classes, but the constructor. We can have a look to see if we can easily move the stuff from the constructor to `connect`, and then backends should not have anything in Client, and we can remove the Client code after the release, without requiring changes to the backends.

jreback commented on ibis/backends/base/sql/alchemy/__init__.py:
```diff
@@ -43,30 +54,369 @@
+        self._inspector.info_cache.clear()
+        return self._inspector
+
+    @staticmethod
```
we typically don't use staticmethods as this is an anti-pattern.

cpcloud replied:
I think this a carryover from the previous location. `staticmethod`s are useful in cases where an otherwise free function is specific to an operation in a class, but doesn't need the instance (`self`) or the type (`cls`) of the class to do its job. I personally think it's fine to leave as a `staticmethod`.

jreback commented on ibis/backends/base/client.py:
```diff
@@ -29,6 +29,127 @@ def exists_database(self, name):
     def exists_table(self, name, database=None):
         return self.backend.exists_table(name, database)
 
+    # Methods below haven't been standardized among backends yet, #3019
```
what is wrong with a Mixin?

cpcloud commented on ibis/backends/base/sql/__init__.py:
```diff
@@ -9,3 +18,200 @@ class BaseSQLBackend(BaseBackend):
+        -------
+        Backend cursor
+        """
+        # TODO results is unused, it can be removed
```
I don't think this comment is valid anymore.

cpcloud commented on ibis/backends/base/sql/__init__.py:
```diff
@@ -9,3 +18,200 @@ class BaseSQLBackend(BaseBackend):
+        # (requires updating Impala tests)
+        # TODO `self.con` is assumed to be defined in subclasses, but there
+        # is nothing that enforces it. We should find a way to make sure
+        # `self.con` is always a DBAPI2 connection, or raise an error
```
What is the precise procedure for determining whether an object is a DBAPI2 connection?

cpcloud commented on ibis/backends/base/sql/__init__.py:
```diff
@@ -9,3 +18,200 @@ class BaseSQLBackend(BaseBackend):
+        """
+        # TODO Reconsider having `kwargs` here. It's needed to support
+        # `external_tables` in clickhouse, but better to deprecate that
+        # feature than all this magic.
```
Can you make an issue for this so that it can be discussed?

cpcloud commented on ibis/backends/base/sql/alchemy/__init__.py:
```diff
@@ -43,30 +54,369 @@
     database_class = AlchemyDatabase
     table_class = AlchemyTable
+    compiler = AlchemyCompiler
+    has_attachment = False
```
What does this mean? Is this referring to the SQLite `ATTACH` statement?

datapythonista replied:
I had a look, this is only used when inserting data in SQLite (`insert` and `load_data`). When `attach` in SQLite has been called, this add the parameter `DataFrame.to_sql(schema=current_database)`.

cpcloud commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -1,370 +1,15 @@
+        self.backend.con = con
+        self.backend.meta = sa.MetaData(bind=con)
+        self.backend._inspector = sa.inspect(con)
+        self.backend._schemas: Dict[str, sch.Schema] = {}
```
We definitely need to remember to go back and make sure we don't have any unnecessary reference cycles once things are moved.

cpcloud commented on ibis/backends/clickhouse/__init__.py:
```diff
@@ -114,3 +128,102 @@ def list_tables(self, like=None, database=None):
+        self.con.disconnect()
+
+    def _fully_qualified_name(self, name, database):
+        if bool(fully_qualified_re.search(name)):
```
This `bool` is unnecessary.

cpcloud commented on ibis/backends/clickhouse/__init__.py:
```diff
@@ -114,3 +128,102 @@ def list_tables(self, like=None, database=None):
+
+    def _ensure_temp_db_exists(self):
+        name = (options.clickhouse.temp_db,)
+        if name not in self.list_databases():
```
Is `list_databases` returning a `list[tuple[str]]`? That seems like it violates the exist contract. Is this just a carryover from the move?

cpcloud commented on ibis/backends/impala/tests/test_patched.py:
```diff
@@ -8,6 +8,7 @@ def patch_execute(con):
 
 
 def test_invalidate_metadata(con, test_data_db):
+    con = con.backend
```
Can you instead return this from the fixture? Not a huge fan of rebinding variables of with different types of values.

datapythonista replied:
Good point, implemented this way first to see if tests were passing, and never updated. Thanks.

cpcloud commented on ibis/backends/mysql/__init__.py:
```diff
@@ -86,3 +91,47 @@ def connect(
+            try:
+                bind.execute("SET @@session.time_zone = 'UTC'")
+            except Exception as e:
+                warnings.warn(f"Couldn't set mysql timezone: {str(e)}")
```
I think this warning is pretty useless, but we can evaluate that again later.

datapythonista replied:
Also moved this from Client without changes, were this was implemented.

cpcloud commented on ibis/backends/sqlite/client.py:
```diff
@@ -1,90 +1,22 @@
-        self.name = path
-        self.database_name = "base"
+        super().__init__(sa.create_engine("sqlite://"))
+        # self.backend.name = path
```
Can this be deleted?

datapythonista replied:
Looks like, thanks for the heads up.

cpcloud commented on ibis/expr/rules.py:
```diff
@@ -291,9 +291,12 @@ def interval(arg, units=None):
 
-    return instance_of(Client, arg)
+    try:
+        return instance_of(Client, arg)
```
I think you can probably write this as:

```python
instance_of((Client, BaseBackend), arg)
```

cpcloud commented on ibis/backends/base/client.py:
```diff
@@ -29,6 +29,127 @@ def exists_database(self, name):
+    def close(self, *args, **kwargs):
+        return self.backend.close(*args, **kwargs)
+
+    # impala
```
Do these methods need to be here? Many of them are highly specific to impala.

cpcloud reviewed (commented):
Looks mostly good, just a few comments.

datapythonista reviewed (commented):
Thanks for the feedback both. If that's fine, I'll leave all the feedback in code that was just moved here to a follow up, and just address the comments on the code implemented or modified here. The diff here is huge, and for any future problem and clearer git history, I think it's better to be able to assume that the big chunks of code here, were moved without changes. And having a separate small PR for the changes, will let revert and identify possible problems in a much easier way.

jreback commented:
ok if this is a straight moved then ok with merging

datapythonista commented:
> ok if this is a straight moved then ok with merging

Yes, all comments not addressed are from code just moved from one class to another. I'll be taking care of those comments in follow up PRs, since it's good feedback that otherwise will get lost.

I addressed every comment in a separate commit, in case it's easier to review, so you don't need to go through the whole diff again. Let me know if there is anything left.

jreback commented:
ok lgtm. prob should have a nice section of the release notes to show the migration path, but that should be a dedicated issue / PR.

jreback commented:
@cpcloud over to you.

cpcloud reviewed (approved):
LGTM.

jreback commented:
thanks @datapythonista
